include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,338 @@
+
+    // perform fast non dominated sort on $$ P_t \cup G_t $$
+    Info << "NSGA2::Optimize() FastNonDominatedSort" << std::endl;
+    ranks.resize(population.size());
+    FastNonDominatedSort(fronts, ranks, calculatedObjectives);
+
+    // perform crowding distance assignment
+    Info << "NSGA2::Optimize() CrowdingDistanceAssignment" << std::endl;
+    crowdingDistance.resize(population.size());
+    for (size_t fNum = 0; fNum < fronts.size(); fNum++)
+    {
+      CrowdingDistanceAssignment(fronts[fNum], objectives, crowdingDistance);
+    }
+
+    // sort based on crowding distance
+    Info << "NSGA2::Optimize() Sort(Crowding Distance)" << std::endl;
+    std::sort(population.begin(),
+              population.end(),
+              [this, ranks, crowdingDistance, population](MatType candidateP,
+                                                          MatType candidateQ)
+              {
+                size_t idxP, idxQ;
+                for(int i = 0; i < population.size(); i++)
+                {
+                  if (arma::approx_equal(population[i], candidateP, "absdiff", epsilon))
+                  {
+                    idxP = i;
+                  }
+                  if (arma::approx_equal(population[i], candidateQ, "absdiff", epsilon))
+                  {
+                    idxQ = i;
+                  }
+                }
+
+                return CrowdingOperator(idxP,
+                                        idxQ,
+                                        ranks,
+                                        crowdingDistance);
+              }
+    );
+
+    // yeild new population P_{t+1}
+    Info << "NSGA2::Optimize() Get P(t+1)" << std::endl;
+    population.resize(populationSize);
+  }
+
+  std::vector<MatType> bestFront;
+
+  for(size_t f: fronts[0])
+  {
+    bestFront.push_back(population[f]);
+  }
+
+  Callback::EndOptimization(*this, objectives, iterate, callbacks...);
+  return bestFront;
+}
+
+template<typename MultiobjectiveFunctionType,
+         typename MatType>
+inline void NSGA2::EvaluateObjectives(std::vector<MatType> population,
+                                      MultiobjectiveFunctionType objectives,
+                                      std::vector<arma::vec>& calculatedObjectives)
+{
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    calculatedObjectives[i] = objectives.Evaluate(population[i]);
+  }
+}
+
+template<typename MatType>
+inline void NSGA2::BinaryTournamentSelection(std::vector<MatType>& population)
+{
+  std::vector<MatType> children;
+
+  while (children.size() < population.size())
+  {
+    size_t indexA = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+    size_t indexB = arma::randi<size_t>(arma::distr_param(0, populationSize - 1));
+
+    if (indexA == indexB)
+    {
+      if (indexB < populationSize - 1)
+        indexB++;
+      else
+        indexB--;
+    }
+
+    MatType parentA = population[indexA];
```

Comment:
I think we can pass those two parameters directly instead of creating another copy first.

Reply:
Okay, will make the changes.